src/r2pipe.rs

```diff
@@ -210,12 +217,42 @@ impl R2Pipe {
 
         // flush out the initial null byte.
         let mut w = [0; 1];
-        sout.read_exact(&mut w)?;
+
+        // if the computer is busy, it might take a bit for the subprocess to connect to
+        // the pipe and we could get back an UnexpectedEof from the pipe in the `read_exact`
+        // call. This will resolve itself when the subprocess fully starts executing and connects
+        // to the pipe, but we don't want to hang indefinitely, so we should put a deadline on
+        // the connecting read that flushes the initial null byte
+        let deadline = Instant::now() + connect_timeout;
+        while Instant::now() < deadline {
+            match sout.read_exact(&mut w) {
+                Ok(_) => break,
+                Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => match child.try_wait()? {
+                    Some(exit_status) => {
+                        return Err(Error::ProgramExited { exit_status });
+                    }
+                    None => {
+                        // rather than just hammering the read_exact call, sleep for a little while
+                        // before trying again
+                        std::thread::sleep(Duration::from_millis(10));
+                        continue;
+                    }
+                },
+                Err(e) => return Err(e.into()),
+            }
+        }
```

Comment:
Rather than handrolling a read timeout (I don't like this arbitrary sleep), is there a primitive that exists already to handle this?

Reply:
This isn't exactly a read timeout. If it were, I actually have a crate that I maintain that would handle that, but that isn't the case here. In this case, the read is immediately returning an error because the read is occurring before the subprocess has attached  to the pipe (its stdout) for writing because the system I am accessing this on is overloaded. The rough timeline is 1) parent process calls fork, 2) parent process tries to read from the pipe, 3) child process finally gets up and going and can write to the pipe. But, we have already lost because step 2) caused and error. If that occurs, what I am trying to do is give the child process on this loaded system a chance to start up and then loop back to try again to clear that initial NUL byte. I could swap the sleep out with a yield, but the reason I put the sleep in place was to force the parent process to give up its time-slice and give the child time to come up. With a yield in place, we will just continue to busy loop while giving the scheduler a chance to prematurely let something else run, but I am not sure that is better. 

What would you prefer?